lmfit/model.py

```diff
@@ -56,6 +56,77 @@ def no_op(*args, **kwargs):
     return no_op
 
 
+def get_reducer(option):
+    """Factory function to build a parser for complex numbers.
+    `option` should be one of `['real', 'imag', 'abs', 'angle']`"""
+    assert option in ['real', 'imag', 'abs', 'angle'], "Unsupported option!"
+
+    def reducer(array):
+            """Convert a complex array to a real array based on the option passed
+            to parse_complex. Does nothing to a purely real array."""
+            if any(np.iscomplex(array)):
+                parsed_array = getattr(np, option)(array)
+            else:
+                parsed_array = array
+
+            return parsed_array
+    return reducer
+
+
+def propagate_err(z, dz, option):
+    """Perform error propagation on a vector of complex uncertainties to
+    get values for magnitude (abs) and phase (angle) uncertainty.
+    Uncertainties are 1/weights. If the weights provided are real, they are
+    assumed to apply equally to the real and imaginary parts. If the weights
+    are complex, the real part of the weights are applied to the real
+    part of the residual and the imaginary part is treated correspondingly."""
+
+    # Check the main vector for complex. Do nothing if real.
+    if any(np.iscomplex(z)):
+        # if uncertainties are real, apply them equally to
+        # real and imaginary parts
+        if all(np.isreal(dz)):
+            dz = dz+1j*dz
+
+        if option == 'real':
+            err = np.real(dz)
+        elif option == 'imag':
+            err = np.imag(dz)
+        elif option in ['abs', 'angle']:
+            rz = np.real(z)
+            iz = np.imag(z)
+
+            rdz = np.real(dz)
+            idz = np.imag(dz)
+
+            # Don't spit out warnings for divide by zero. Will fix these later.
+            with np.errstate(divide='ignore', invalid='ignore'):
+
+                if option == 'abs':
+                    # Standard error propagation for abs = sqrt(re**2 + im**2)
+                    err = np.true_divide(np.sqrt((iz*idz)**2+(rz*rdz)**2),
+                                         np.abs(z))
+
+                    # For abs = 0, error is +/- abs(rdz + j idz)
+                    err[err == np.inf] = np.abs(dz)[err == np.inf]
+
+                if option == 'angle':
+                    # Standard error propagation for angle = arctan(im/re)
+                    err = np.true_divide(np.sqrt((rz*idz)**2+(iz*rdz)**2),
+                                         np.abs(z))
+
+                    # For abs = 0, error is +/- pi (i.e. the whole circle)
+                    err[err == np.inf] = np.pi
+
+        else:
+            # Should never make it here, but don't want things to break
```

Comment:
I guess if this should never happen, you either don't need to define an ```else``` block if you're sure about it. Alternatively, you probably want to throw an exception - I am not sure whether setting the error to ```0.0``` is a good idea in this case.

Reply:
Yeah, this is a good point. I've added an assert to check the value passed to option and also raised a ValueError here. Hopefully this will future proof it for when someone down the row decides to "upgrade" it!